File: 2017/test/day_13_test.exs
defmodule AdventOfCode2017.Day13Test do
  use ExUnit.Case

  import AdventOfCode2017.Day13
  import AdventOfCode2017.TestHelper

  test "part1" do
    assert part1(sample_input_1()) == 24
    assert part1(input("test/input/day_13.txt")) == 1624
  end

  test "part2" do
    assert part2(sample_input_1()) == 10
    assert part2(input("test/input/day_13.txt")) == 3923436
  end

  test "parse" do
    assert parse(sample_input_1()) == sample_layers_1()
  end

  defp sample_input_1, do: "0: 3\n1: 2\n4: 4\n6: 4"

  defp sample_layers_1, do: [{0, 3}, {1, 2}, {4, 4}, {6, 4}]
end

File: 2018/test/day_13_test.rb
require "test_helper"
require "day_13"

class Day13Test < Minitest::Test
  def setup
    @day_13 = Day13.new
  end

  def test_part1
    assert_equal "0,3", @day_13.part1(sample_input_1)
    assert_equal "7,3", @day_13.part1(sample_input_2)
    assert_equal "102,114", @day_13.part1(input_no_strip("day_13.txt"))
  end

  def test_part2
    assert_equal "6,4", @day_13.part2(sample_input_3)
    assert_equal "146,87", @day_13.part2(input_no_strip("day_13.txt"))
  end

  def test_parse
    _, tracks = @day_13.parse(sample_input_1)
    _, tracks2 = @day_13.parse(sample_input_2)

    assert_equal [["|"], ["|"], ["|"], ["|"], ["|"], ["|"], ["|"]], tracks
    assert_equal [["/", "-", "-", "-", "\\"], ["|", " ", " ", " ", "|", " ", " ", "/", "-", "-", "-", "-", "\\"], ["|", " ", "/", "-", "+", "-", "-", "+", "-", "\\", " ", " ", "|"], ["|", " ", "|", " ", "|", " ", " ", "|", " ", "|", " ", " ", "|"], ["\\", "-", "+", "-", "/", " ", " ", "\\", "-", "+", "-", "-", "/"], [" ", " ", "\\", "-", "-", "-", "-", "-", "-", "/"]], tracks2
  end

  def sample_input_1
    "|
v
|
|
|
^
|"
  end

  def sample_input_2
    %q(/->-\
|   |  /----\
| /-+--+-\  |
| | |  | v  |
\-+-/  \-+--/
  \------/)
  end

  def sample_input_3
    %q(/>-<\
|   |
| /<+-\
| | | v
\>+</ |
  |   ^
  \<->/)
  end
end
